Add initial docs for config-deployment CM (#2841)
knative/docs: docs/serving/services/deployment.md

---
title: "Deployment Configuration"
linkTitle: "Deployment Configuration"
weight: 03
type: "docs"
---

The `config-deployment` ConfigMap is located in the `knative-serving` namespace.
This ConfigMap, known as the Deployment ConfigMap, contains settings that determine how Kubernetes `Deployment` resources, that back Knative services, are configured.

## Accessing the Deployment ConfigMap 

To view the current Deployment ConfigMap:

```bash
kubectl get configmap -n knative-serving config-deployment -oyaml
```

## Progress Deadline Seconds

> ProgressDeadline setting specifies the value of the `.spec.progressDeadlineSeconds` field on the underlying Kubernetes deployment for the Knative Revision.
> By default this value is set to 120 seconds.
> The value is expressed as a Go `time.Duration` string representation, but must be rounded to a second precision.
> Please see [Kubernetes documentation](https://kubernetes.io/docs/concepts/workloads/controllers/deployment/#progress-deadline-seconds) for details about this setting.

Reviewer: ```suggestion
The `.spec.progressDeadlineSeconds` setting specifies the number of seconds you want to wait for your Deployment to progress before the system reports back that the Deployment has failed progressing for the Knative Revision.
By default, this value is set to 120 seconds.
The value is expressed as a Go `time.Duration` string representation, but must be rounded to a second precision.
See the [Kubernetes documentation](https://kubernetes.io/docs/concepts/workloads/controllers/deployment/#progress-deadline-seconds) for more information.
```

> In Knative the Autoscaler component gives up and scales the revision to 0 if the initial scale cannot be achieved within approximately this time span. If this happens the revision is considered to be failed.

Reviewer: ```suggestion
The Knative Autoscaler component scales the revision to 0 if the initial scale value cannot be achieved within the time limit defined by `.spec.progressDeadlineSeconds`. If this occurs, the revision deployment is considered to have failed.
```

Reviewer: Not sure if this is accurate, but if not can you please clarify what "failing" means in this context? Is there some "failed" error message or status users can look at somewhere to figure out if the revision failed?

Author: So the `.spec.progressDeadlineSeconds` is not relevant here (it's deployment spec), here we should actually refer to the configmap setting we're describing how to set — since it's all in purview of knative.

Failing here means any problems that don't lead us to scaling to the initial scale.

In the following cases it might make sense to change this setting to a higher value:

- it takes a lot of time to pull the Service image (e.g. they are very big) 
- it takes a lot of time to make the service ready (e.g. priming initial cache state)

> - the cluster is relying on cluster autoscaler to allocate resources for new pods.

Reviewer: I'd delete this and move it higher in the doc as per suggestion comment above

Author: Commited the other change.

The following is an incomplete snippet of a ConfigMap that sets this value to 10 minutes.

```yaml
apiVersion: v1
kind: ConfigMap
metadata:
  name: config-deployment
  namespace: knative-serving
data:
...
  progressDeadline: "10m"
...

> ```

Reviewer: I'd rewrite this part as its own equal section - either change the "## Progress Deadline Seconds" heading to just "## Progress Deadline" and then talk about seconds and minutes in the intro and show both examples, or add a separate section for "## Progress Deadline Minutes" :slightly_smiling_face: lmk if this isn't clear or you need help with suggestions. Thanks!

Author: added a sub heading.

---

> title: "Deployment Configuration"

Reviewer: ```suggestion
title: "Modifying the Deployment Config Map"
```

Reviewer: I'd maybe update the heading actually to this and remove the linkTitle?

linkTitle: "Deployment Configuration"
weight: 03
type: "docs"
---

> The Deployment ConfigMap contains settings that configure how Kubernetes Deployment resource, backing Knative Services, is configured.

Reviewer: ```suggestion
The `config-deployment` ConfigMap contains settings that determine how Kubernetes `Deployment` resources, that back Knative services, are configured.
```

## Accessing the Deployment ConfigMap 

> The Deployment ConfigMap is named `config-deployment` and is located in the `knative-serving` namespace.

Reviewer: Remove this line considering above comment? :slightly_smiling_face: just a suggestion for readability.

To view the current Deployment ConfigMap:

```bash
kubectl get configmap -n knative-serving config-deployment -oyaml
```

> ## progressDeadline

Reviewer: ```suggestion
## Progress Deadline Seconds
```

> ## Deployment Progress Deadline Setting
> 
> The `.spec.progressDeadlineSeconds` setting specifies the number of seconds you want to wait for your Deployment to progress before the system reports back that the Deployment has failed progressing for the Knative Revision. By default, this value is set to 120 seconds.
> The value is expressed as a Go `time.Duration` string representation, but must be rounded to a second precision.
> The Knative Autoscaler component scales the revision to 0 and overall Knative service enters terminal `Failed` state, if the initial scale cannot be achieved within the time limit defined by this setting.
> 
> You may want to configure this setting as a higher value if any of the following issues occur in your Knative deployment:
> 
> - It takes a long time to pull the Service image, due to the size of the image.
> - It takes a long time for the Service to become `READY`, due to priming of the initial cache state.
> - The cluster is relies on cluster autoscaling to allocate resources for new pods.
> 
> See the [Kubernetes documentation](https://kubernetes.io/docs/concepts/workloads/controllers/deployment/#progress-deadline-seconds) for more information.
> 
> The following is an incomplete snippet of a ConfigMap that sets this value to 10 minutes.
> 
> ### Configuring the Deployment Progress Deadline setting value
> 
> ```yaml
> apiVersion: v1
> kind: ConfigMap
> metadata:
>   name: config-deployment
>   namespace: knative-serving
> data:
> ...
>   progressDeadline: "10m"
> ...
> ```

Reviewer: ```suggestion
## Configuring progress deadlines

Configuring progress deadline settings allows you to specify the maximum time, either in seconds or minutes, that you will wait for your Deployment to progress before the system reports back that the Deployment has failed progressing for the Knative Revision.
By default, this value is set to 120 seconds.
The value is expressed as a Go `time.Duration` string representation, but must be rounded to a second precision.

The Knative Autoscaler component scales the revision to 0, and the Knative service enters a terminal `Failed` state, if the initial scale cannot be achieved within the time limit defined by this setting.

You may want to configure this setting as a higher value if any of the following issues occur in your Knative deployment:

- It takes a long time to pull the Service image, due to the size of the image.
- It takes a long time for the Service to become `READY`, due to priming of the initial cache state.
- The cluster is relies on cluster autoscaling to allocate resources for new pods.

See the [Kubernetes documentation](https://kubernetes.io/docs/concepts/workloads/controllers/deployment/#progress-deadline-seconds) for more information.

The following example shows a snippet of an example Deployment Config Map that sets this value to 10 minutes:

```yaml
apiVersion: v1
kind: ConfigMap
metadata:
  name: config-deployment
  namespace: knative-serving
data:
...
  progressDeadline: "10m"
...
```
```

Reviewer: Suggested editing, sorry for all the reviews!
